### PR TITLE
Add SARIF support

### PR DESCRIPTION
**Pull request by shaopeng-gh**
Please see discussion thread: https://github.com/rrrene/credo-proposals/issues/58

@michaelcfanning @EasyRhinoMSFT @rrrene

### CODE REVIEW

**Comment by rrrene** [comment]
Thanks for putting this together :+1:

File: lib/credo/cli/output/formatter/sarif.ex [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+      to_file_uri(exec.cli_options.path)
+    )
+    |> print_map()
+  end
+
+  def print_map(map) do
+    map
+    |> prepare_for_json()
+    |> Jason.encode!(pretty: true)
+    |> move_version_to_top()
```
[the remark this answers is not in the document]
Comment (EasyRhinoMSFT) [reply lib/credo/cli/output/formatter/sarif.ex:177]: This allows consumers (such as the SARIF Viewer for Visual Studio) to determine the SARIF version of the log without having to read the entire file. The version is important to consumers who might take some action: the SARIF Viewer, for instance, offers to perform a migration to the latest version of the SARIF specification.

File: lib/credo/cli/output/formatter/sarif.ex [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        "id" => Credo.Code.Name.full(issue.check),
+        "fullDescription" => %{
+          "text" => issue.check.explanation
+        },
+        "properties" => %{
+          "tags" => [
+            issue.category
+          ]
+        },
+        "helpUri" => "https://hexdocs.pm/credo/#{Credo.Code.Name.full(issue.check)}.html"
```
[the remark this answers is not in the document]
Comment (rrrene) [reply lib/credo/cli/output/formatter/sarif.ex]: > I just pushed an update to `master` which allows you to call

Tests are failing now. Did you rebase your branch against `master`?

[the remark this answers is not in the document]
Comment (shaopeng-gh) [reply lib/credo/cli/output/formatter/sarif.ex]: Found the fixed the issue, no need of () in the `"helpUri" => issue.check.docs_uri()`.

**Comment by shaopeng-gh** [comment]
I have a few improvements need to make per Michael's input, will be back with a new commit:
1. In the result node, remove the redundant level: warning. 
2. The description of the checks looks like is in markdown format, in SARFI `message` have a node `markdown` to put it in. 
This will be for the rendering tools that can render markdown.
And in `text` field we will also set it, but replacing the `
This will be for the rendering tools that can not render markdown or choose to render the text field.

**Comment by shaopeng-gh** [comment]
fyi see a possible issue, noted down seperatly here: https://github.com/rrrene/credo/issues/1017

**Comment by michaelcfanning** [comment]
@shaopeng-gh, just checking with you, are we code complete with this proposal? i.e., you have implemented all the changes we discussed offline? Really appreciate your work her, this will make a difference to Elixir devs, particularly those interacting with GHAS. 

@osfernandez

**Comment by shaopeng-gh** [comment]
> @shaopeng-gh, just checking with you, are we code complete with this proposal? i.e., you have implemented all the changes we discussed offline? Really appreciate your work her, this will make a difference to Elixir devs, particularly those interacting with GHAS.
> 
> @osfernandez

Yes Michael. The last change was @rrrene kindly provided rule id and rule name and I have used it in the SARIF. 
@rrrene let me know if you can approve the pr, thanks,

**Comment by michaelcfanning** [comment]
That's great! This is very nice work. I know for a fact there are several eager customers out there who plan to use this new reporting capability. :)

**Comment by rrrene** [comment]
This is part of Credo [v1.7.0-rc.1](https://hex.pm/packages/credo/1.7.0-rc.1) :+1:

**Pull request merged**